openssl-dynamic/src/main/java/io/netty/tcnative/jni/SSL.java

```diff
@@ -249,28 +256,71 @@ private SSL() { }
+    public static native int writeToBIO(long bioAddress, long wbufAddress, int wlen);
 
     /**
-     * BIO_read
-     * @param bio
-     * @param rbuf
-     * @param rlen
-     * @return
+     * Initialize the BIO for the SSL instance. This is a custom BIO which is designed to play nicely with a direct
+     * {@link java.nio.ByteBuffer}. Because it is a special BIO it requires special usage such that
+     * {@link #setBIOBuffer(long, long, int, boolean)} and {@link #clearBIOBuffer(long)} are called in order to provide
+     * to supply data to SSL, and also to ensure the internal SSL buffering mechanism is expecting write at the appropriate times.
+     *
+     * @see #initBufferBIO(long, int)
+     */
+    public static long initBufferBIO(long ssl) {
+        return initBufferBIO(ssl, 0);
+    }
+
+    /**
+     * Initialize the BIO for the SSL instance. This is a custom BIO which is designed to play nicely with a direct
+     * {@link java.nio.ByteBuffer}. Because it is a special BIO it requires special usage such that
+     * {@link #setBIOBuffer(long, long, int, boolean)} and {@link #clearBIOBuffer(long)} are called in order to provide
+     * to supply data to SSL, and also to ensure the internal SSL buffering mechanism is expecting write at the appropriate times.
+     *
+     * @param ssl the SSL instance (SSL *)
+     * @param internalWriteBIOSize The maximum size of the internal BIO used to buffer write operations that may not
+     *                             be the direct result of the application attempting to encrypt data. Pass 0 to use the
+     *                             default max size.
+     * @return pointer to the Network BIO (BIO *).
+     *         The memory is owned by {@code ssl} and will be cleaned up by {@link #freeSSL(long)}.
+     */
+    public static native long initBufferBIO(long ssl, int internalWriteBIOSize);
+
+    /**
+     * Set the memory location which that OpenSSL's internal BIO will use to write encrypted data to, or read encrypted
+     * data from.
+     * <p>
+     * After you are done buffering data you should call {@link #clearBIOBuffer(long)}.
+     * @param bio {@code BIO*}.
+     * @param bufferAddress The memory address (typically from a direct {@link java.nio.ByteBuffer}) which will be used
+     *                    to either write encrypted data to, or read encrypted data from by OpenSSL's internal BIO pair.
+     * @param maxUsableBytes The maximum usable length in bytes starting at {@code bufferAddress}.
+     * @param isSSLWriteSink {@code true} if this buffer is expected to buffer data as a result of calls to {@code SSL_write}.
+     *                       {@code false} if this buffer is expected to buffer data as a result of calls to {@code SSL_read}.
      */
-    public static native int readFromBIO(long bio, long rbuf, int rlen);
+    public static native void setBIOBuffer(long bio, long bufferAddress, int maxUsableBytes, boolean isSSLWriteSink);
```

Comment:
@Scottmitch just a nit but should we be more consistent with naming. Like sometimes we start with `bio` sometimes we have `BIO` as suffix etc.